### PR TITLE
Add historical averaging

### PR DESCRIPTION
**Pull request by Aniket1998**
Add support for Historical Averaging of Weights as discussed in Salimans et. al. 2016

### CODE REVIEW

File: torchgan/losses/historical.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            else:
+                optimizer_generator.zero_grad()
+                loss = None
+                for i, p in enumerate(generator.parameters()):
+                    if loss is None:
+                        loss = torch.sum((p - self.sum_parameters[i]) ** 2)
+                    else:
+                        loss += torch.sum((p - self.sum_parameters[i]) ** 2)
+                    self.sum_parameters[i] += p
+                loss /= self.timesteps
```
Comment (avik-pal) [inline torchgan/losses/historical.py]: Is this what the paper suggests?
You should be taking the average of the past parameters,i.e `self.sum_parameters[i] / self.timesteps`. You might divide the loss by the length of `generator.parameters` for normalization purposes.

File: torchgan/losses/historical.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                self.timesteps += 1
+                return 0.0
+            else:
+                optimizer_discriminator.zero_grad()
+                loss = None
+                for i, p in enumerate(discriminator.parameters()):
+                    if loss is None:
+                        loss = (p - self.sum_parameters[i]).norm(2)
+                    else:
+                        loss += (p - self.sum_parameters[i]).norm(2)
```
Comment (avik-pal) [inline torchgan/losses/historical.py]: We should maintain consistency by using `norm` or `torch.sum(...) ** 2`.

File: torchgan/losses/historical.py (hunk 1 above)
Comment (Aniket1998) [reply torchgan/losses/historical.py]: Fixed in the new commit, was a typo I overlooked

File: torchgan/losses/historical.py (hunk 2 above)
Comment (Aniket1998) [reply torchgan/losses/historical.py]: Fixed in the new commit, torch.sum is more appropriate.

**Comment by Aniket1998** [comment]
@avik-pal fixed the issue, ran some error checks on my end and it runs fine now. Could get a small example tested on MNIST (just add the Historical Average losses to the DCGAN Minimax example in the Model zoo) and then merge accordingly?

**Comment by avik-pal** [comment]
Yes on it.

**Comment by avik-pal** [comment]
@Aniket1998 Model is not converging. The generated images have barely changed since the first epoch. I am using Minimax + Historical Averaging on both generator and discriminator. Already trained for 25 epochs on MNIST. Adam with 0.001 lr for both.
